### PR TITLE
Add counteract experiment atb allocation

### PR DESCRIPTION
**Pull request by nataliebanegas**
Add counteract experiment atb allocation

**Reviewer:**
@sammacbeth @sballesteros 

## Description:
Allocates users who installed the extension but haven't been assigned any atb into the counteract experiment.

## Steps to test this PR:
1. `npm run dev-chrome`
2. In Chrome, install the extension (the dev version), wait for extension-success page to open, verify that either `atb=v388-2vk` or `atb=v388-2vj` is present in the URL.

## Automated tests:
- [x] Unit tests
- [x] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/atb.js [hunk 1]
```diff
@@ -161,6 +168,67 @@ const ATB = (() => {
...
+         * The minorVersion rolls over every night at midnight ET
+         *
+         * @typedef {Object} AtbVersionInfo
+         * @property {number} majorVersion - # of weeks since noon EST on 2/24/16
+         * @property {number} minorVersion - # of days into the current week
+         * @property {string} version - the version string combining both major and minor version. Format: `v${majorVersion}-${minorVersion}`
+         *
+         * @return {AtbVersionInfo}
+         */
+        getATBVersionInfo: () => {
```
Comment (nataliebanegas) [inline shared/js/background/atb.js:183]: I've ported `getATBVersionInfo` from the static-pages repo as I didn't find it in the extension codebase.

File: shared/js/background/atb.js [hunk 2]
```diff
@@ -161,6 +168,67 @@ const ATB = (() => {
...
+            return {
+                minorVersion: minorVersion,
+                majorVersion: majorVersion,
+                version: 'v' + majorVersion + '-' + minorVersion
+            }
+        },
+
+        getChromeCounteractExpATB: (atbString) => {
+            const variant = 'v'
+            const atbVariant = Math.random() < 0.5 ? 'k' : 'k'
```
Comment (jonathanKingston) [inline shared/js/background/atb.js:228]: Are both left and right supposed to be the same here?

File: shared/js/background/atb.js [hunk 3]
```diff
@@ -161,6 +168,67 @@ const ATB = (() => {
...
+            const epoch = isDST ? ATB_EPOCH - ONE_HOUR : ATB_EPOCH
+
+            // time in ms since DST adjusted epoch:
+            const timeSinceATBEpoch = localTime - epoch
+
+            const majorVersion = Math.ceil(timeSinceATBEpoch / ONE_WEEK)
+            const minorVersion = Math.ceil((timeSinceATBEpoch % ONE_WEEK) / ONE_DAY)
+
+            return {
+                minorVersion: minorVersion,
```
Comment (jonathanKingston) [inline shared/js/background/atb.js:220]: The lint is expecting:
```suggestion
                minorVersion,
```

File: shared/js/background/atb.js [hunk 4]
```diff
@@ -161,6 +168,67 @@ const ATB = (() => {
...
+                majorVersion: majorVersion,
```
Comment (jonathanKingston) [inline shared/js/background/atb.js:221]: Same here:
```suggestion
                majorVersion,
```

File: shared/js/background/atb.js (hunk 1 above)
Comment (jonathanKingston) [reply shared/js/background/atb.js:183]: We *currently* just pull it from the server instead using the tab URLs

File: shared/js/background/atb.js (hunk 2 above)
Comment (nataliebanegas) [reply shared/js/background/atb.js]: nice catch @jonathanKingston, this was done for testing only - updated the letters to two variants.

File: shared/js/background/atb.js [hunk 5]
```diff
@@ -175,6 +243,11 @@ const ATB = (() => {
                         return !!atb
                     })
 
+                    // in case there is no assigned atb, enroll into Chrome Counteract experiment
```
Comment (jonathanKingston) [inline shared/js/background/atb.js:246]: Should we be filtering out non chrome users?

File: shared/js/background/atb.js [hunk 6]
```diff
@@ -175,6 +243,11 @@ const ATB = (() => {
...
+                    if (!atb) {
+                        atb = ATB.getChromeCounteractExpATB(ATB.getATBVersionInfo().version)
```
Comment (sammacbeth) [inline shared/js/background/atb.js:248]: I don't think we need the `getATBVersionInfo` implementation here, as `ATB.setInitialVersions` will already have pulled a plain ATB version from the server which you can get with `settings.getSetting('atb')`. Then it's just a case of appending the experiment.

File: shared/js/background/atb.js [hunk 7]
```diff
@@ -175,6 +243,11 @@ const ATB = (() => {
                         return !!atb
                     })
 
+                    // in case there is no assigned atb, enroll into Chrome Counteract experiment
+                    if (!atb) {
```
Comment (sammacbeth) [inline shared/js/background/atb.js:247]: As Jonathon mentioned, you probably want to filter this to chrome only.
```suggestion
                    if (!atb && utils.getBrowserName() === 'chrome') {
```

**Review by sammacbeth — CHANGES_REQUESTED** [review]
I think we should drop the `getATBVersionInfo` code, as we should rely on the server for this value. Doing this on the client is error prone, as the user clock/timezone might not be correct.

File: shared/js/background/atb.js (hunk 5 above)
Comment (nataliebanegas) [reply shared/js/background/atb.js]: yes, updated!

File: shared/js/background/atb.js [hunk 8]
```diff
@@ -175,6 +182,11 @@ const ATB = (() => {
                         return !!atb
                     })
 
+                    // in case there is no assigned atb variant, enroll into Chrome Counteract experiment
+                    if (!atb && settings.getSetting(atb) && settings.getSetting(atb).atb && utils.getBrowserName() === 'chrome') {
+                        atb = ATB.getChromeCounteractExpATB(settings.getSetting(atb).atb)
```
Comment (nataliebanegas) [inline shared/js/background/atb.js:187]: I'd like to confirm that `settings.getSetting(atb)` will return a value, even in the case when extension is installed directly from the Chrome store

File: shared/js/background/atb.js [hunk 9]
```diff
@@ -175,6 +182,11 @@ const ATB = (() => {
                         return !!atb
                     })
 
+                    // in case there is no assigned atb variant, enroll into Chrome Counteract experiment
+                    if (!atb && settings.getSetting(atb) && settings.getSetting(atb).atb && utils.getBrowserName() === 'chrome') {
```
Comment (sammacbeth) [inline shared/js/background/atb.js:186]:
```suggestion
                    if (!atb && settings.getSetting('atb') && utils.getBrowserName() === 'chrome') {
```
You were calling `settings.getSetting(undefined)` which actually returns an object with all settings values. You should instead just ask specifically for the 'atb' setting: settings.getSetting('atb')

File: shared/js/background/atb.js (hunk 8 above)
Comment (sammacbeth) [inline shared/js/background/atb.js:187]:
```suggestion
                        atb = ATB.getChromeCounteractExpATB(settings.getSetting('atb'))
```
As above: just pull the 'atb' setting

File: shared/js/background/atb.js (hunk 9 above)
Comment (nataliebanegas) [reply shared/js/background/atb.js]: ugh, my mistake, should be fixed now

**Comment by kimmillet** [comment]
P

Sent from my iPhone
All nana is great

On 28 Jul,2023, at 4:41 PM, Natalie Banegas ***@***.***> wrote:

﻿
@nataliebanegas commented on this pull request. In shared/js/background/atb.js: > @@ -175,6 +182,11 @@ const ATB = (() => { return !!atb }) + // in case there is no assigned atb variant, enroll into C
<https://duckduckgo.com/-hbnVH5CveyAR5Ci_-BN0aFP5V1aj8JAmvWxCLMPFEcBDiwdpye3Jyit4eO_BLaaX8DC1tVdrm-r-4TTWVj9APbzna79x4bGf1PkIQEXLAFvzS9XfDFAdOkIoCnvl91NQY7iZO5S3YNTaRJL9f2VM2OPzRa4cJhFwuXe9_Y9dxmGuTN-Bzfth5g0-wCuKRFynuWrlDlxvbSZcb6bxbUT9x2lt23scL1r2hisRyE4d3KfpdaNQN5Zrit89BX36whAAn0aL0FWpSkyhXWtV8Y3LhRphkbt333QgAhWSBKLGvwKmM-Eufc7Mcw4cxsLDNjCkXmF9aoOxzOOUOw5z2n0gx5GfxE6we02W2oFE8_Z3T8H9d1jbJgjgx80RDo4FMgZdtkSFAzQdHHcGvNLNL28TxlDA14b>
DuckDuckGo removed 1 tracker.
More →

@nataliebanegas commented on this pull request.

________________________________

In shared/js/background/atb.js<https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/2145#discussion_r1278066659>:

> @@ -175,6 +182,11 @@ const ATB = (() => {
                         return !!atb
                     })

+                    // in case there is no assigned atb variant, enroll into Chrome Counteract experiment
+                    if (!atb && settings.getSetting(atb) && settings.getSetting(atb).atb && utils.getBrowserName() === 'chrome') {


ugh, my mistake, should be fixed now

**Review by sammacbeth — APPROVED** [review]

**Review by jonathanKingston — APPROVED** [review]

File: shared/js/background/atb.js [hunk 10]
```diff
@@ -175,6 +182,11 @@ const ATB = (() => {
                         return !!atb
                     })
 
+                    // in case there is no assigned atb variant, enroll into Chrome Counteract experiment
+                    if (!atb && settings.getSetting('atb') && utils.getBrowserName() === 'chrome') {
```
Comment (moollaza) [inline shared/js/background/atb.js:186]: I think we want to limit to EN locale only?

**Pull request merged**